servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright © 2022 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Publisher.defer;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.concurrent.internal.ConcurrentUtils.releaseLock;
+import static io.servicetalk.concurrent.internal.ConcurrentUtils.tryAcquireLock;
+import static io.servicetalk.utils.internal.PlatformDependent.newUnboundedSpscQueue;
+import static java.lang.Math.min;
+
+final class PublisherFlatMapConcatUtils {
+    private PublisherFlatMapConcatUtils() {
+    }
+
+    static <T, R> Publisher<R> flatMapConcatSingle(final Publisher<T> publisher,
+                                                   final Function<? super T, ? extends Single<? extends R>> mapper) {
+        return defer(() -> {
+            final Queue<Item<R>> results = newUnboundedSpscQueue(4);
+            final AtomicInteger consumerLock = new AtomicInteger();
+            return publisher.flatMapMergeSingle(orderedMapper(mapper, results, consumerLock))
+                    .shareContextOnSubscribe();
+        });
+    }
+
+    static <T, R> Publisher<R> flatMapConcatSingleDelayError(
+            final Publisher<T> publisher, final Function<? super T, ? extends Single<? extends R>> mapper) {
+        return defer(() -> {
+            final Queue<Item<R>> results = newUnboundedSpscQueue(4);
+            final AtomicInteger consumerLock = new AtomicInteger();
+            return publisher.flatMapMergeSingleDelayError(orderedMapper(mapper, results, consumerLock))
+                    .shareContextOnSubscribe();
+        });
+    }
+
+    static <T, R> Publisher<R> flatMapConcatSingle(final Publisher<T> publisher,
+                                                   final Function<? super T, ? extends Single<? extends R>> mapper,
+                                                   final int maxConcurrency) {
+        return defer(() -> {
+            final Queue<Item<R>> results = newUnboundedSpscQueue(min(8, maxConcurrency));
```

Comment:
this lower bounds is not mentioned in the Publisher docs.

Reply:
The queue is unbound, the parameter is just an estimate for initial size of the queue. This value is not appropriate for javadocs as it is not part of the API contract and can be adjusted if necessary.